### PR TITLE
feat: Add Caching for ElementUtil

### PR DESCRIPTION
**Pull request by lqiu96**
Fixes #[126](https://github.com/googleapis/java-docfx-doclet/issues/126)

Running the kokoro script with the updated element caching (no concurrency)
image
image
image

Seems the jobs without concurrency takes ~90 minutes. This is without running mvn clean install (which may take ~30 minutes/ ~12 minutes with -T 1C). Since the mvn javadoc:aggregate job needs to run mvn compile, do we need to have the have it `clean install`?

I don't think we would run into any memory issues. I can try to reduce the initial capacity since we're no longer running this for the entire monorepo (only for each submodule in the monorepo).

Both these changes were part of the initial debugging with YourKit profiler. The concurrent code calling getEnclosedElements() would need to be synchronized between the different threads. They were making the same call to element.getEnclosedElements() (https://docs.oracle.com/javase/7/docs/api/javax/lang/model/element/Element.html#getEnclosedElements()) but the result for this function doesn't change. Adding a map/cache to store the already computed values.

- [x] Tests pass
- [x] Appropriate changes to README are included in PR

### CODE REVIEW

File: third_party/docfx-doclet-143274/src/main/java/com/microsoft/build/Lookup.java [hunk 1]
```diff
@@ -12,14 +12,17 @@
 
 public class Lookup {
 
-    private Map<String, String> globalLookup = new HashMap<>();
-    private Map<String, Map<String, String>> localLookupByFileName = new HashMap<>();
+    private static final int INITIAL_CAPACITY = 10000;
```
Comment (lqiu96) [inline third_party/docfx-doclet-143274/src/main/java/com/microsoft/build/Lookup.java:15]: Set large Map capacity to *try* and reduce the number of copy/ increases needed.

File: third_party/docfx-doclet-143274/src/main/java/com/microsoft/lookup/BaseLookup.java [hunk 2]
```diff
@@ -32,6 +33,7 @@
 
 public abstract class BaseLookup<T extends Element> {
 
+    private static final int INITIAL_CAPACITY = 500000;
```
Comment (lqiu96) [inline third_party/docfx-doclet-143274/src/main/java/com/microsoft/lookup/BaseLookup.java:36]: Set large initial map capacity to *try* and reduce the number of copy/ increases needed. This is for ClassLookup, ClassItemsLookup, and PackageLookup. Maybe I can allow configuration of this capacity for the different lookups. ClassLookup (mid size), ClassItemsLookup (large), PackageLookup (small)

File: third_party/docfx-doclet-143274/src/main/java/com/microsoft/util/ElementUtil.java [hunk 3]
```diff
@@ -15,34 +17,54 @@
 
 public class ElementUtil {
 
+    private static final int INITIAL_CAPACITY = 1000000;
```
Comment (lqiu96) [inline third_party/docfx-doclet-143274/src/main/java/com/microsoft/util/ElementUtil.java:20]: Set large initial map capacity to *try* and reduce the number of copy/ increases needed.

File: third_party/docfx-doclet-143274/src/main/java/com/microsoft/util/ElementUtil.java [hunk 4]
```diff
@@ -15,34 +17,54 @@
...
+        return ElementFilter.typesIn(getEnclosedElements(element)).stream()
+            .filter(o -> !Utils.isPrivateOrPackagePrivate(o))
+            .filter(o -> !matchAnyPattern(excludeClasses, String.valueOf(o.getQualifiedName())))
+            .sorted((o1, o2) ->
+                StringUtils.compare(String.valueOf(o1.getSimpleName()), String.valueOf(o2.getSimpleName()))
+            ).collect(Collectors.toList());
+    }
+
+    public List<? extends Element> getEnclosedElements(Element element) {
+        elementMap.computeIfAbsent(element, this::getEnclosedElementsInternal);
```
Comment (suztomo) [inline third_party/docfx-doclet-143274/src/main/java/com/microsoft/util/ElementUtil.java:54]: My memo: this is the cache.

**Comment by suztomo** [comment]
For my future learning, would you enrich this pull request description with your observations that explain why you focused on optimizing `[getEnclosedElements()](https://docs.oracle.com/en/java/javase/11/docs/api/java.compiler/javax/lang/model/element/Element.html#getEnclosedElements())`?

**Review by suztomo — APPROVED** [review]

**Comment by lqiu96** [comment]
@suztomo Sure. Changes were initially part this draft PR: https://github.com/googleapis/java-docfx-doclet/pull/139, which was two parts 1. Add concurrency and 2. Add Element caching. When I created the original PR, I came across a test failure with the concurrency code so I split out the caching part while I debug the other issue.

Both these changes were part of the initial debugging with YourKit profiler. The concurrent code calling getEnclosedElements() would need to be synchronized between the different threads. They were making the same call to `element.getEnclosedElements()` (https://docs.oracle.com/javase/7/docs/api/javax/lang/model/element/Element.html#getEnclosedElements()) but the result for this function doesn't change. Adding a map/cache to store the already computed values.

**Comment by suztomo** [comment]
@lqiu96 Thanks. Would you paste here the screenshot of YourKit profiler showing getEnclosedElements?
(If you didn't take it, then never mind)

**Comment by lqiu96** [comment]
@suztomo I didn't take a screenshot unfortunately.

**Pull request merged**